File: Ash/Mode/7_pcsemaphore.py
import threading
import time
from threading import Semaphore

bs = 8

es = Semaphore(bs)
fs = Semaphore(0)
buffer = [0] * bs
in_index = 0
out_index = 0

def producer():
    global in_index
    item = 1

    while True:
        print(f"Produced item: {item}")

        es.acquire()

        buffer[in_index] = item
        in_index = (in_index + 1) % bs

        fs.release()

        item += 1
        time.sleep(1)  # Added sleep to simulate time taken to produce an item

def consumer():
    global out_index

    while True:
        fs.acquire()

        consumed = buffer[out_index]
        print(f"Consumed item: {consumed}")

        out_index = (out_index + 1) % bs

        es.release()
        time.sleep(1)  # Added sleep to simulate time taken to consume an item

if __name__ == "__main__":
    producer_thread = threading.Thread(target=producer)
    consumer_thread = threading.Thread(target=consumer)

    producer_thread.start()
    consumer_thread.start()

    producer_thread.join()
    consumer_thread.join()

File: Ash/OS_LAB/7_pcsemaphore.c
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>

#define bs 8

sem_t es;
sem_t fs;
int buffer[bs];
int in = 0, out = 0;

void *producer(void *arg)
{
    int item = 1;

    while (1)
    {

        printf("Produced item: %d\n", item);

        sem_wait(&es);

        buffer[in] = item;
        in = (in + 1) % bs;

        sem_post(&fs);

        item++;
    }

    pthread_exit(NULL);
}

void *consumer(void *arg)
{
    while (1)
    {

        sem_wait(&fs);

        int consumed = buffer[out];
        printf("Consumed item: %d\n", consumed);

        out = (out + 1) % bs;

        sem_post(&es);
    }

    pthread_exit(NULL);
}

int main()
{
    pthread_t producerThread, consumerThread;
    sem_init(&es, 0, bs);
    sem_init(&fs, 0, 0);

    pthread_create(&producerThread, NULL, producer, NULL);
    pthread_create(&consumerThread, NULL, consumer, NULL);

    pthread_join(producerThread, NULL);
    pthread_join(consumerThread, NULL);

    sem_destroy(&es);
    sem_destroy(&fs);

    return 0;
}